pywt/_thresholding.py

```diff
@@ -245,3 +248,77 @@ def threshold_firm(data, value_low, value_high):
     if np.any(large_vals[0]):
         thresholded[large_vals] = data[large_vals]
     return thresholded
+
+
+def estimate_sigma(data):
+    """
+    Robust wavelet-based estimator of the (Gaussian) noise standard deviation.
+    Parameters
+    ----------
+    data : ndarray
+        The data used to estimate sigma.
+
+    Returns
+    -------
+    sigma : float
+        Estimated noise standard deviation.
+
+    Notes
+    -----
+    This function assumes the noise follows a Gaussian distribution. The
+    estimation algorithm is based on the median absolute deviation of the
+    wavelet detail coefficients as described in section 4.2 of [1]_.
+
+    References
+    ----------
+    .. [1] D. L. Donoho and I. M. Johnstone. "Ideal spatial adaptation
+       by wavelet shrinkage." Biometrika 81.3 (1994): 425-455.
+       DOI:10.1093/biomet/81.3.425
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> import pywt
+    >>> data = np.sin(np.linspace(0,10,1000))
+    >>> np.random.seed(42)
+    >>> noise = np.random.normal(0,1,1000)
+    >>> sigma = pywt.estimate_sigma(data + noise)
+    """
+
+    coeffs = dwtn(data, wavelet='db2')
+    detail_coeffs = coeffs['d' * data.ndim]
+    return _sigma_est_dwt(detail_coeffs, distribution='Gaussian')
+
+
+def _sigma_est_dwt(detail_coeffs, distribution='Gaussian'):
+    """Calculate the robust median estimator of the noise standard deviation.
+    Parameters
+    ----------
+    detail_coeffs : ndarray
+        The detail coefficients corresponding to the discrete wavelet
+        transform of an image.
+    distribution : str
+        The underlying noise distribution.
+    Returns
+    -------
+    sigma : float
+        The estimated noise standard deviation (see section 4.2 of [1]_).
+    References
+    ----------
+    .. [1] D. L. Donoho and I. M. Johnstone. "Ideal spatial adaptation
+       by wavelet shrinkage." Biometrika 81.3 (1994): 425-455.
+       DOI:10.1093/biomet/81.3.425
+    """
+    # Consider regions with detail coefficients exactly zero to be masked out
+    detail_coeffs = detail_coeffs[np.nonzero(detail_coeffs)]
+
+    if distribution.lower() == 'gaussian':
+        # 75th quantile of the underlying, symmetric noise distribution
+        # denom = scipy.stats.norm.ppf(0.75)
```

Comment:
In addition to a string name, you could accept any object with a `ppf` method here perhaps (assuming that works, didn't read the paper)? Just check with `if hasattr(distribution, 'ppf')`.

The bigger issue here seems to be the hardcoding of the 75th quantile. Why is that not a keyword?

Reply:
It seems like both choices originate from specific suggestions by the author, I agree that the best solution would be to have any distribution or quantile as valid inputs, as that could prove useful. The issue I saw is the use of scipy.stats in the original code, because PyWavelets doesn't rely on scipy (right?) I'm not sure what the best thing to do there is, including scipy for just this is overkill, perhaps a couple hand coded options for now? Or for just Gaussian with a variable percent input, shouldn't be too hard to get going without scipy.

Reply:
Btw, nice suggestion for finding if a distribution supports a method, seeing the example it seems obvious now.

Reply:
Indeed we don't want to rely on `scipy`. That's why I suggested checking for a `ppf` method; users can still pass in `scipy.stats` distributions but we don't need to import from scipy anywhere so we don't rely on scipy.

Reply:
A concise explanation of the rationale for use of the 75th quantile in MAD is given on the following wikipedia page:
https://en.wikipedia.org/wiki/Median_absolute_deviation
Perhaps it is worth adding that to the References? I don't remember if it was spelled out explicitly in Donoho's paper

I also agree about not relying on `scipy` here. `scipy` was already a dependency of `scikit-image` and a reviewer of my PR there preferred to call the `ppf` method explicitly instead of hard-coding the value, but that doesn't mean we have to do the same here.

The idea to check for a `ppf` method is nice.

Another approach to deal with other noise distributions such as Poisson noise is to first apply a [variance stabilizing transform](https://en.wikipedia.org/wiki/Variance-stabilizing_transformation) prior to calling this Gaussian-based MAD method. An example implementation of this type of approach for use with Rician noise is available [here](http://www.cs.tut.fi/~foi/RiceOptVST/), but it is not under compatible license terms.